File: apps/android/app/src/main/java/com/commerceos/android/viewmodel/AddressPlatformState.kt
package com.commerceos.android.viewmodel

import com.commerceos.android.location.GeoPoint
import com.commerceos.android.location.GeocodedPlace
import com.commerceos.android.location.PlaceSearchResult
import com.commerceos.android.model.AddressConflictWarning
import com.commerceos.android.model.ApiAddress
import com.commerceos.android.model.StructuredAddress

/**
 * Production State Machine steps for Commerce OS Address & Location Platform.
 * ViewModel is the SOLE source of truth for flow navigation.
 */
sealed class AddressPlatformStep {
    /** Browsing saved addresses in address book */
    object AddressBook : AddressPlatformStep()

    /** Autocomplete place/location search */
    data class SearchingLocation(val query: String = "") : AddressPlatformStep()

    /** Location selected on real map, pin positioned */
    data class LocationSelected(val point: GeoPoint) : AddressPlatformStep()

    /** Async debounced reverse geocoding in progress */
    data class ReverseGeocoding(val point: GeoPoint) : AddressPlatformStep()

    /** Trust boundary: Location pin confirmation step */
    data class ConfirmingPin(val place: GeocodedPlace) : AddressPlatformStep()

    /** Structured address details form */
    object EditingDetails : AddressPlatformStep()
}

/**
 * Explicit Save State for Address persistence operations.
 */
sealed class SaveState {
    object Idle : SaveState()
    object Saving : SaveState()
    data class Success(val savedAddress: ApiAddress) : SaveState()
    data class Error(val message: String) : SaveState()
}

/**
 * Explicit GPS acquisition state.
 */
sealed class LocationAcquisitionState {
    object Idle : LocationAcquisitionState()
    object AcquiringGps : LocationAcquisitionState()
    data class Success(val point: GeoPoint) : LocationAcquisitionState()
    data class Error(val message: String) : LocationAcquisitionState()
}

/**
 * Single Canonical UI State Container for Address & Location Platform.
 * Maintains explicit separation between the immutable original edit snapshot
 * and the user's mutable draft address.
 */
data class AddressPlatformUiState(
    val currentStep: AddressPlatformStep = AddressPlatformStep.AddressBook,
    val originalAddress: StructuredAddress? = null,
    val draftAddress: StructuredAddress = StructuredAddress(),
    val isEditingExisting: Boolean = false,
    val editingAddressId: String? = null,
    val locationSearchQuery: String = "",
    val placeSearchResults: List<PlaceSearchResult> = emptyList(),
    val isSearchingPlaces: Boolean = false,
    val activeGeocodedPlace: GeocodedPlace? = null,
    val isReverseGeocoding: Boolean = false,
    val locationAcquisitionState: LocationAcquisitionState = LocationAcquisitionState.Idle,
    val saveState: SaveState = SaveState.Idle,
    val addressConflict: AddressConflictWarning = AddressConflictWarning(hasConflict = false),
    val locationErrorMessage: String? = null
) {
    val isFlowActive: Boolean
        get() = currentStep !is AddressPlatformStep.AddressBook

    val isDirty: Boolean
        get() = originalAddress != null && draftAddress != originalAddress

    val activeFormAddress: StructuredAddress
        get() = draftAddress
}

File: apps/ios/CommerceOS/Features/Address/Models/AddressPlatformState.swift
import Foundation

/**
 * Production State Machine steps for Commerce OS Address & Location Platform.
 * ViewModel is the SOLE source of truth for flow navigation.
 * 1:1 Parity with Android AddressPlatformStep in AddressPlatformState.kt
 */
public enum AddressPlatformStep: Equatable {
    /** Browsing saved addresses in address book */
    case addressBook

    /** Autocomplete place/location search */
    case searchingLocation(query: String)

    /** Location selected on real map, pin positioned */
    case locationSelected(point: GeoPoint)

    /** Async debounced reverse geocoding in progress */
    case reverseGeocoding(point: GeoPoint)

    /** Trust boundary: Location pin confirmation step */
    case confirmingPin(place: GeocodedPlace)

    /** Structured address details form */
    case editingDetails
}

/**
 * Explicit Save State for Address persistence operations.
 * 1:1 Parity with Android SaveState in AddressPlatformState.kt
 */
public enum SaveState: Equatable {
    case idle
    case saving
    case success(AddressDto)
    case error(String)
}

/**
 * Explicit GPS acquisition state.
 * 1:1 Parity with Android LocationAcquisitionState in AddressPlatformState.kt
 */
public enum LocationAcquisitionState: Equatable {
    case idle
    case acquiringGps
    case success(GeoPoint)
    case error(String)
}

/**
 * Single Canonical UI State Container for Address & Location Platform.
 * Maintains explicit separation between the immutable original edit snapshot
 * and the user's mutable draft address.
 * 1:1 Parity with Android AddressPlatformUiState in AddressPlatformState.kt
 */
public struct AddressPlatformUiState: Equatable {
    public var currentStep: AddressPlatformStep
    public var originalAddress: StructuredAddress?
    public var draftAddress: StructuredAddress
    public var isEditingExisting: Bool
    public var editingAddressId: String?
    public var locationSearchQuery: String
    public var placeSearchResults: [PlaceSearchResult]
    public var isSearchingPlaces: Bool
    public var activeGeocodedPlace: GeocodedPlace?
    public var isReverseGeocoding: Bool
    public var locationAcquisitionState: LocationAcquisitionState
    public var saveState: SaveState
    public var addressConflict: AddressConflictWarning
    public var locationErrorMessage: String?

    public init(
        currentStep: AddressPlatformStep = .addressBook,
        originalAddress: StructuredAddress? = nil,
        draftAddress: StructuredAddress = StructuredAddress(),
        isEditingExisting: Bool = false,
        editingAddressId: String? = nil,
        locationSearchQuery: String = "",
        placeSearchResults: [PlaceSearchResult] = [],
        isSearchingPlaces: Bool = false,
        activeGeocodedPlace: GeocodedPlace? = nil,
        isReverseGeocoding: Bool = false,
        locationAcquisitionState: LocationAcquisitionState = .idle,
        saveState: SaveState = .idle,
        addressConflict: AddressConflictWarning = AddressConflictWarning(hasConflict: false),
        locationErrorMessage: String? = nil
    ) {
        self.currentStep = currentStep
        self.originalAddress = originalAddress
        self.draftAddress = draftAddress
        self.isEditingExisting = isEditingExisting
        self.editingAddressId = editingAddressId
        self.locationSearchQuery = locationSearchQuery
        self.placeSearchResults = placeSearchResults
        self.isSearchingPlaces = isSearchingPlaces
        self.activeGeocodedPlace = activeGeocodedPlace
        self.isReverseGeocoding = isReverseGeocoding
        self.locationAcquisitionState = locationAcquisitionState
        self.saveState = saveState
        self.addressConflict = addressConflict
        self.locationErrorMessage = locationErrorMessage
    }

    public var isFlowActive: Bool {
        currentStep != .addressBook
    }

    public var isDirty: Bool {
        originalAddress != nil && draftAddress != originalAddress
    }

    public var activeFormAddress: StructuredAddress {
        draftAddress
    }
}
